### PR TITLE
Tempo Sync Modulation

### PR DESCRIPTION
**Pull request by BuffMcBigHuge**
# Tempo Sync & Modulation

## Summary

Adds tempo synchronization and beat-synced parameter modulation to Scope. Pipelines can lock to Ableton Link or MIDI clock, quantize discrete parameter changes to beat boundaries, and continuously modulate parameters (e.g. `noise_scale`, `denoising_steps`) in sync with the beat. Target users are VDJs running Scope alongside Ableton Live, Resolume, or similar applications.

### Core

- **Tempo sync** — Ableton Link and MIDI clock support with optional dependencies (`aalink`, `mido`/`python-rtmidi`). When neither is installed, the UI shows install hints and the feature degrades gracefully.
- **Parameter scheduling** — Discrete parameter changes (prompt switches, denoising adjustments) can be quantized to beat/bar boundaries with configurable lookahead to compensate for pipeline latency.
- **Parameter modulation** — Continuous per-frame modulation of numeric parameters driven by beat phase. Pipeline-agnostic: operates on kwargs before they reach any pipeline.

### Frontend

- **Tempo Sync panel** — Source toggle (Link/MIDI), BPM input, beats per bar, live beat indicator, quantize mode, lookahead, beat cache reset rate, prompt cycle rate.
- **Modulation section** — Target selector (from schema `ui.modulatable` fields), wave shape (sine, cosine, triangle, saw, square, pulse decay), rate (½ beat → 4 bars), depth.
- **Prompt cycling** — Automatically advance through timeline prompts on beat/bar boundaries when enabled.

### Pipelines

- **metronome** — Visual metronome for testing ParameterScheduler lookahead. Three color layers, `latency_ms` to simulate pipeline delay.
- **mod-scope** — Oscilloscope for modulation visualization.

### Modulatable pipelines

LongLive, Krea, MemFlow, RewardForcing, StreamDiffusionV2 expose `denoising_steps` and (LongLive) `noise_scale` as modulatable via `ui_field_config(modulatable=True, ...)`.

---

## Architecture

```
Tempo Sources (Link / MIDI / Client-forwarded)
    → TempoSync Manager (BeatState, 15Hz notification loop)
    → ParameterScheduler (quantize discrete changes)
    → ModulationEngine (per-frame param modulation)
    → PipelineProcessor (inject beat state + modulated params)
    → Pipeline
```

- **Beat state:** `bpm`, `beat_phase`, `bar_position`, `beat_count`, `is_playing` injected into every pipeline call.
- **Cloud mode:** Browser forwards beat state over WebRTC data channel; server prefers client state when fresh (< 2s).
- **Modulation:** Config sent as `modulations` in parameter updates; Pydantic-validated on receipt.

---

## REST API

| Method | Path | Purpose |
|--------|------|---------|
| GET | `/api/v1/tempo/status` | Current tempo status and beat state |
| POST | `/api/v1/tempo/enable` | Enable tempo sync (Link or MIDI) |
| POST | `/api/v1/tempo/disable` | Disable tempo sync |
| POST | `/api/v1/tempo/set_tempo` | Set BPM (Link only) |
| GET | `/api/v1/tempo/sources` | Available tempo sources and capabilities |

**Modulation:** No dedicated REST API. Config is sent as `modulations` in WebRTC parameter updates.

---

## How to try it

1. **Install optional deps** (for Link or MIDI):
   ```bash
   uv sync --group link   # Ableton Link
   uv sync --group midi   # MIDI clock
   ```

2. **Run Scope** and start a stream.

3. **Open the Tempo Sync panel** in the sidebar. Enable tempo sync, choose Link or MIDI, set BPM.

4. **Quantize mode:** Set to "Beat" or "Bar" so changes like prompt switches land on the beat. Adjust lookahead if latency is visible.

5. **Modulation:** In the modulation section, pick a target (e.g. Noise Scale), choose shape/rate/depth, and toggle ON.

6. **Prompt cycling:** Add multiple prompts to the timeline, set `promptCycleRate` to "Bar" or "Beat", and prompts will advance automatically.

7. **Demo pipelines:** Use `metronome` to test lookahead or `mod-scope` to visualize modulation.

---

## Testing

- **ParameterScheduler:** `tests/test_parameter_scheduler.py` — adversarial tests for boundary math, concurrency, lookahead, and edge cases.
- **E2E:** Tests passed against fal.ai preview deployment.

---

## Optional dependencies

```toml
[project.optional-dependencies]
link = ["aalink>=0.1.1"]
midi = ["mido>=1.3.0", "python-rtmidi>=1.5.0"]
```

---

## Breaking changes

None. Tempo sync and modulation are additive. Pipelines that don't use beat kwargs are unaffected.

---

## Documentation

- `docs/tempo-sync-review.md` — Implementation review, architecture, API, edge cases, and future work.

---

## Summary by CodeRabbit

* **New Features**
  * Tempo synchronization via Ableton Link and MIDI Clock support
  * Beat-synced parameter modulation with waveform shapes, depth, and rate controls
  * Parameter scheduling aligned to beat boundaries with quantization modes
  * New Metronome and Modulation Scope visualization pipelines for testing
  * Beat-driven prompt cycling

* **Bug Fixes**
  * Enhanced noise controller state initialization handling

### CODE REVIEW

File: src/scope/core/pipelines/longlive/schema.py [hunk 1]
```diff
@@ -102,7 +102,12 @@ class LongLiveConfig(BasePipelineConfig):
         default=[1000, 750, 500, 250],
         description="Denoising step schedule for progressive generation",
         json_schema_extra=ui_field_config(
-            order=6, component="denoising_steps", is_load_param=True
+            order=6,
+            component="denoising_steps",
```
Comment (leszko) [inline src/scope/core/pipelines/longlive/schema.py:106]: Stupid question: so the tempo sync always changes/modulates _only_ the `denoising_steps` param, right? It's not something a user can map/select?

File: src/scope/core/pipelines/mod_scope/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    "U": [0b101, 0b101, 0b101, 0b101, 0b010],
+    "V": [0b101, 0b101, 0b101, 0b101, 0b010],
+    "X": [0b101, 0b101, 0b010, 0b101, 0b101],
+    "_": [0b000, 0b000, 0b000, 0b000, 0b111],
+}
+
+ALL_PATTERNS = {**_DIGIT_PATTERNS, **_ALPHA_PATTERNS}
+
+TRACE_LENGTH = 192
+BEATS_PER_BAR = 4
```
Comment (leszko) [inline src/scope/core/pipelines/mod_scope/pipeline.py:78]: Shouldn't it be a param that we set in UI?

File: src/scope/core/pipelines/mod_scope/schema.py
Comment (leszko) [inline src/scope/core/pipelines/mod_scope/schema.py:1]: Do I understand correctly that both `metronome` and `mod_scope` are sample pipelines to present the usage of tempo sync? If yes, then maybe we need just 1 "hello-world" pipeline like that, wdyt?

File: src/scope/core/pipelines/wan2_1/blocks/noise_scale_controller.py [hunk 3]
```diff
@@ -119,11 +119,15 @@ def __call__(self, components, state: PipelineState) -> tuple[Any, PipelineState
...
-            block_state.init_cache = True
+            # Cache init on discrete jumps from the motion controller
+            if (
+                block_state.manage_cache
+                and block_state.current_noise_scale is not None
+                and block_state.current_noise_scale != block_state.noise_scale
+            ):
+                block_state.init_cache = True
+        # When noise_controller is off (e.g. during modulation), let noise_scale
+        # changes flow through without triggering expensive cache resets.
```
Comment (leszko) [inline src/scope/core/pipelines/wan2_1/blocks/noise_scale_controller.py:130]: Why do we need this change? The modulation takes care of the denoising step list, so why do we change anything about the noise controller?

File: src/scope/server/webrtc.py [hunk 4]
```diff
@@ -349,10 +367,21 @@ def on_data_channel_message(message):
...
 
+                        # Check for quantized update flag
+                        if data.pop("_quantized", False):
+                            if session.video_track and hasattr(
+                                session.video_track, "frame_processor"
+                            ):
+                                session.video_track.frame_processor.schedule_quantized_update(
+                                    data
+                                )
+                            return
```
Comment (leszko) [inline src/scope/server/webrtc.py:383]: I guess this will not work with the cloud. I think we could make it working with the cloud, but I wonder if the tempo sync feature even makes sense in the cloud? Won't the latency make it unusable?

File: src/scope/server/tempo_router.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+# ---------------------------------------------------------------------------
+
+
+def _get_tempo_sync() -> "TempoSync | None":
+    from .app import tempo_sync
+
+    return tempo_sync
+
+
+@router.get("/status", response_model=TempoStatusResponse)
```
Comment (leszko) [inline src/scope/server/tempo_router.py:38]: This will not proxied to the cloud. I _guess_ it's correct. But won't to confirm.

File: src/scope/core/pipelines/mod_scope/pipeline.py (hunk 2 above)
Comment (BuffMcBigHuge) [reply src/scope/core/pipelines/mod_scope/pipeline.py:78]: Considering that mod_scope and metronome are simple tests for tempo sync, should we just remove them entirely @leszko ?

Comment (leszko) [reply src/scope/core/pipelines/mod_scope/pipeline.py:78]: We can remove and move to plugins. Or if you prefer, we can keep one. But probably 2 "hello worlds" tests is too much for the scope core.

File: src/scope/core/pipelines/mod_scope/schema.py
Comment (BuffMcBigHuge) [reply src/scope/core/pipelines/mod_scope/schema.py:1]: I think it's best that we remove both.

File: src/scope/core/pipelines/longlive/schema.py (hunk 1 above)
Comment (BuffMcBigHuge) [reply src/scope/core/pipelines/longlive/schema.py:106]: The system is schema-driven and user-selectable.

Backend pipeline schemas opt in per-field using modulatable=True in ui_field_config(), along with optional modulatable_min/modulatable_max bounds.

This way plugins can opt-in!

File: src/scope/core/pipelines/wan2_1/blocks/noise_scale_controller.py (hunk 3 above)
Comment (BuffMcBigHuge) [reply src/scope/core/pipelines/wan2_1/blocks/noise_scale_controller.py:130]: This change is needed for noise_scale modulation (not just denoising steps). LongLive exposes noise_scale and denoising_steps as modulatable. I'm not necessarily happy with noise modulation, it works - but it's not very impressive. I'm sure it can be tuned better.

**Comment by BuffMcBigHuge** [comment]
Closing this PR in lieu of https://github.com/daydreamlive/scope/pull/703.
